VDF.Core/Utils/GrayBytesUtils.cs

```diff
@@ -68,38 +67,235 @@ static class GrayBytesUtils {
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private static unsafe float PercentageDifferenceAvx2(byte[] img1, byte[] img2) {
+			long diff = 0;
+			Vector256<ushort> vec = Vector256<ushort>.Zero;
+			Span<Vector256<byte>> vImg1 = MemoryMarshal.Cast<byte, Vector256<byte>>(img1);
+			Span<Vector256<byte>> vImg2 = MemoryMarshal.Cast<byte, Vector256<byte>>(img2);
+
+			for (int i = 0; i < vImg1.Length; i++)
+				vec = Avx2.Add(vec, Avx2.SumAbsoluteDifferences(vImg2[i], vImg1[i]));
+
+			for (int i = 0; i < Vector256<ushort>.Count; i++)
+				diff += Math.Abs(vec.GetElement(i));
+
 			return (float)diff /  img1.Length / 256;
 		}
 
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private static unsafe float PercentageDifferenceSse2(byte[] img1, byte[] img2) {
+			long diff = 0;
+			Vector128<ushort> vec = Vector128<ushort>.Zero;
+			Span<Vector128<byte>> vImg1 = MemoryMarshal.Cast<byte, Vector128<byte>>(img1);
+			Span<Vector128<byte>> vImg2 = MemoryMarshal.Cast<byte, Vector128<byte>>(img2);
+
+			for (int i = 0; i < vImg1.Length; i++)
+				vec = Sse2.Add(vec, Sse2.SumAbsoluteDifferences(vImg2[i], vImg1[i]));
+
+			for (int i = 0; i < Vector128<ushort>.Count; i++)
+				diff += Math.Abs(vec.GetElement(i));
+
+			return (float)diff /  img1.Length / 256;
+		}
+
+
+	// Different FlipGrayScale versions:
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private static byte[] FlipGrayScaleReverse(byte[] img, int img_width)
+		{
+			int rows = img.Length / img_width;
+			byte[] flip_img = (byte[])img.Clone();
+			for (int r = 0; r < rows; r++)
+				Array.Reverse(flip_img, r*img_width, img_width);
+			return flip_img;
+		}
+
+		private static byte[] flipp_shuf256 = {
+				15,14,13,12,11,10, 9, 8,   7, 6, 5, 4, 3, 2, 1, 0, 
+				31,30,29,28,27,26,25,24,  23,22,21,20,19,18,17,16 
+		};
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private static byte[] FlipGrayScaleAvx2(byte[] img)
+		{
+			byte[] flip_img = new byte[img.Length];
+			Span<Vector256<byte>> vImg = MemoryMarshal.Cast<byte, Vector256<byte>>(img);
+			Span<Vector256<byte>> vImg_flipped = MemoryMarshal.Cast<byte, Vector256<byte>>(flip_img);
+			Span<Vector256<byte>> vFlipp_shuf = MemoryMarshal.Cast<byte, Vector256<byte>>(flipp_shuf256);
+
+			for (int i = 0; i < vImg.Length; i++)
+				vImg_flipped[i] = Avx2.Shuffle(vImg[i], vFlipp_shuf[0]);
+
+			return flip_img;
+		}
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		private static byte[] FlipGrayScaleSse2(byte[] img)
+		{
+			byte[] flip_img = new byte[img.Length];
+			Span<Vector128<byte>> vImg = MemoryMarshal.Cast<byte, Vector128<byte>>(img);
+			Span<Vector128<byte>> vImg_flipped = MemoryMarshal.Cast<byte, Vector128<byte>>(flip_img);
+			Span<Vector128<byte>> vFlipp_shuf = MemoryMarshal.Cast<byte, Vector128<byte>>(flipp_shuf256);
+
+			for (int i = 0; i < vImg.Length; i++)
+				vImg_flipped[i] = Avx2.Shuffle(vImg[i], vFlipp_shuf[0]);
+
+			return flip_img;
+		}
+
+#if false
```

Comment:
Remove this

Reply:
Do you have a place for unit / performance / ... tests?
Yes, the file is not the best place, but I didn't want to keep the code section from you, in case you also wanted to try out the variants in comparison.

Reply:
I dont have a place yet. Its still on my very long to do list.

Reply:
Removed